tools/wpt/run.py

```diff
@@ -302,36 +364,61 @@ def setup_kwargs(self, kwargs):
         if not kwargs["device_serial"]:
             kwargs["device_serial"] = ["emulator-5554"]
 
+        if kwargs["webdriver_binary"] is None and "wdspec" in kwargs["test_types"]:
+            webdriver_binary = None
+            if not kwargs["install_webdriver"]:
+                webdriver_binary = self.browser.find_webdriver()
+
+            if webdriver_binary is None:
+                install = self.prompt_install("geckodriver")
+
+                if install:
+                    logger.info("Downloading geckodriver")
+                    webdriver_binary = self.browser.install_webdriver(
+                        dest=self.venv.bin_path,
+                        channel=kwargs["browser_channel"],
+                        browser_binary=kwargs["binary"])
+            else:
+                logger.info("Using webdriver binary %s" % webdriver_binary)
+
+            if webdriver_binary:
+                kwargs["webdriver_binary"] = webdriver_binary
+            else:
+                logger.info("Unable to find or install geckodriver, skipping wdspec tests")
+                kwargs["test_types"].remove("wdspec")
+
         for device_serial in kwargs["device_serial"]:
             if device_serial.startswith("emulator-"):
                 # We're running on an emulator so ensure that's set up
-                emulator = android.install(logger,
-                                           reinstall=False,
-                                           no_prompt=not self.prompt,
-                                           device_serial=device_serial)
                 android.start(logger,
```

Comment:
In Chromium we are starting Android emulator and install the necessary packages before we start Wptrunner. If we are going to run Chrome Android at upstream also, we should have a way to not starting Android emulator and installing packages.

This is needed anyway, because when developers are running tests locally, they tend to launch the emulator themselves, and keep it running, and update the package whenever needed. Those things are all slow, so give developers an option to control that by themselves is useful.

Reply:
I believe that if there's already an emulator running this will do the right thing, but I haven't actually checked that. In any case if it doesn't I'd be happy to take later patches to ensure that we don't do unnecessary work in the case that a device is already configured.

Reply:
Ack